client/concordclient/src/concord_client.cpp

```diff
@@ -143,13 +143,20 @@ void ConcordClient::createGrpcConnections() {
 }
 
 void ConcordClient::checkAndReConnectGrpcConnections() {
+  bool need_reconnection = false;
   for (size_t con_offset = 0; con_offset < config_.topology.replicas.size(); ++con_offset) {
     if (!(grpc_connections_[con_offset])->isConnected()) {
+      need_reconnection = true;
+      break;
+    }
+  }
+  if (need_reconnection) {
+    for (size_t con_offset = 0; con_offset < config_.topology.replicas.size(); ++con_offset) {
       auto trsc_config = std::make_unique<GrpcConnectionConfig>(config_.subscribe_config.use_tls,
                                                                 config_.subscribe_config.pem_private_key,
                                                                 config_.subscribe_config.pem_cert_chain,
                                                                 config_.transport.event_pem_certs);
-      (grpc_connections_[con_offset])->reConnect(trsc_config);
+      (grpc_connections_[con_offset])->checkAndReConnect(trsc_config);
```

Comment:
If connection at a `con_offset` is not connected, we reconnect to all the replicas? Why don't we reconnect to the replica which was disconnected?

Reply:
Thanks, @shgandhi, we don't reconnect all replicas, we only reconnect that replica that is disconnected. That's why we check again within the write lock before disconnecting. If it's not disconnected, then inside checkAndReConnect we don't reconnect.

Another question will be why do we have 2 loops?
Ans: The first loop is for isConnected() function which takes a read lock and quickly determines whether a reconnection is needed or not. This happens 99% of the time. But sometimes we actually need to reconnect. But when we reconnect we should check and then only reconnect as it might be possible that some other thread got the write lock and reconnect. That's why we have 2 loops and we check always before reconnection, so we essentially never reconnect any already connected replica.